lib/autsr.gi

```diff
@@ -939,6 +945,198 @@ local ff,r,d,ser,u,v,i,j,k,p,bd,e,gens,lhom,M,N,hom,Q,Mim,q,ocr,split,MPcgs,
 
 end);
 
+
+BindGlobal("AGSRModuleLayerSeries",function(g)
+local s,l,r,i,j,sy,hom,p,pcgs;
+  s:=ShallowCopy(DerivedSeriesOfGroup(g));
+  l:=s[Length(s)];
+  r:=RadicalGroup(s[Length(s)]);
+  if Size(r)>1 then # cannot be last, as solvable
+    Append(s,DerivedSeriesOfGroup(r));
+  fi;
+  i:=2;
+  while i<=Length(s) do
+    if HasAbelianFactorGroup(s[i-1],s[i]) then
+      p:=Factors(IndexNC(s[i-1],s[i]))[1];
+      # is it a single prime?
+      if not IsPrimePowerInt(IndexNC(s[i-1],s[i])) then
+        hom:=NaturalHomomorphismByNormalSubgroupNC(s[i-1],s[i]);
+        sy:=SylowSystem(Image(hom));
+        l:=[sy[1]];
+        for j in [2..Length(sy)-1] do
+          Add(l,ClosureGroup(l[Length(l)],sy[j]));
+        od;
+        l:=Reversed(l);
+        l:=List(l,x->PreImage(hom,x));
+        Info(InfoMorph,6,"insert prime @",i);
+        s:=Concatenation(s{[1..i-1]},l,s{[i..Length(s)]});
+      elif not HasElementaryAbelianFactorGroup(s[i-1],s[i]) then
+        # not elementary abelian -- pth powers suffice as abelian
+        l:=s[i];
+        for j in GeneratorsOfGroup(s[i-1]) do
+          l:=ClosureGroup(l,j^p);
+        od;
+        Info(InfoMorph,6,"insert ppower @",i);
+        s:=Concatenation(s{[1..i-1]},[l],s{[i..Length(s)]});
+      else
+        # make module
+        pcgs:=ModuloPcgs(s[i-1],s[i]);
+        l:=LinearActionLayer(g,pcgs);
+        l:=GModuleByMats(l,GF(p));
+        # check for characteristic submodules
+        r:=MTX.BasisRadical(l);
+        if Length(r)=0 then
+          r:=MTX.BasisSocle(l);
+          if Length(r)=l.dimension then
+            # semisimple -- use homogeneous
+            r:=List(MTX.CollectedFactors(l),x->x[1]);
+            if Length(r)>1 then
+              r:=MTX.Homomorphisms(l,r[1]);
+              Error("hom");
```

Comment:
Is the `Error` here because this should never happen? If so, perhaps you could expand the error message to give a bigger hint about why something has gone wrong.

Reply:
Sorry, I see you changed this in the subsequent commit.